auth_token/contrib/is_core_auth/views.py

```diff
@@ -1,18 +1,72 @@
+from django import forms
+from django.contrib.auth import login as auth_login
+from django.contrib.auth.hashers import make_password
+from django.contrib.auth.views import LoginView as DjangoLoginView
+from django.core.exceptions import ImproperlyConfigured
+from django.forms.utils import ErrorList
+from django.http import HttpResponseRedirect
+from django.utils.translation import ugettext as _
 from django.views.generic.base import RedirectView
 
+import import_string
+from auth_token.config import settings
+from auth_token.contrib.common.views import LoginView as _LoginView
+from auth_token.contrib.common.views import LogoutView as _LogoutView
+from auth_token.contrib.is_core_auth.forms import LoginCodeVerificationForm
+from auth_token.models import Token
+from auth_token.utils import generate_two_factor_code, login, takeover
 from is_core.generic_views import DefaultCoreViewMixin
 from is_core.generic_views.mixins import GetCoreObjViewMixin
 
-from auth_token.config import settings
-from auth_token.contrib.common.views import LoginView as _LoginView, LogoutView as _LogoutView
-from auth_token.utils import takeover
-
 
 class LoginView(_LoginView):
 
     template_name = 'is_core/login.html'
 
 
+class TwoFactorLoginView(LoginView):
+
+    def _generate_and_send_two_factor_code(self):
+        code = generate_two_factor_code(Token.TWO_FACTOR_CODE_LENGTH)
+        import_string(settings.AUTHENTICATION_METHOD)(self.request.token, code)
+        self.request.token.two_factor_code = make_password(code, salt=Token.TWO_FACTOR_CODE_SALT)
+        self.request.token.save()
+
+    def _login(self, user, expiration, form):
+        login(
+            self.request, user, expiration, allowed_cookie=self.allowed_cookie, allowed_header=self.allowed_header,
+            two_factor_login=True
+        )
+
+    def get_success_url(self):
+        # send user the key for second part of authentication process
+        self._generate_and_send_two_factor_code()
+        return '{url}?{redirect_field_name}={value}'.format(
+            url=settings.TWO_FACTOR_REDIRECT_URL, redirect_field_name=self.redirect_field_name,
+            value=self.get_redirect_url()
+        )
+
+    def form_valid(self, form):
+        """
+        The user has provided valid credentials (this was checked in AuthenticationForm.is_valid()). So now we
+        can check the test cookie stuff and log him in.
+        """
+        self._login(form.get_user(), not form.is_permanent(), form)
+
+        if not hasattr(self.request.token.user, settings.TWO_FACTOR_USER_CONTACT_FIELD_NAME):
+            raise ImproperlyConfigured(_('User model does not have {} field.').format(
+                settings.TWO_FACTOR_USER_CONTACT_FIELD_NAME)
+            )
+        elif not getattr(self.request.token.user, settings.TWO_FACTOR_USER_CONTACT_FIELD_NAME):
```

Comment:
This might throw `AttributeError`.

Reply:
maybe we can have custom exceptions which function for sending code can throw. According to type of exception the ImproperlyConfigured or some validation error will be show to the client